### PR TITLE
HIVE-25065: Implement ALTER TABLE for setting iceberg table properties

### PR DESCRIPTION
**Pull request by lcspinter**
### What changes were proposed in this pull request?
When running ALTER TABLE ... SET/UNSET TBLPROPERTIES on an Iceberg table the properties should be updated on the Iceberg side as well. 

### Why are the changes needed?
The collection of changed property keys are collected on `AbstractAlterTableOperation` and stored in `EnvironemntContext`. `HiveIcebergMetaHook#commitAlterTable` takes care of the processing of `EnvironmentContext` and updating of the iceberg table. 

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Manual test and unit test

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java [hunk 1]
```diff
@@ -136,6 +137,22 @@ private void finalizeAlterTableWithWriteIdOp(Table table, Table oldTable, List<P
...
     try {
       environmentContext.putToProperties(HiveMetaHook.ALTER_TABLE_OPERATION_TYPE, desc.getType().name());
+      if (desc.getType() == AlterTableType.ADDPROPS) {
+        Map<String, String> oldTableParameters = oldTable.getParameters();
+        environmentContext.putToProperties(HiveMetaHook.SET_PROPERTIES,
+            table.getParameters().entrySet().stream()
+                .filter(e -> !oldTableParameters.containsKey(e.getKey()) ||
+                    !oldTableParameters.get(e.getKey()).equals(e.getValue()))
+                .map(Map.Entry::getKey)
+                .collect(Collectors.joining(HiveMetaHook.PROPERTIES_SEPARATOR)));
```
Comment (pvary) [inline ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:147]: Does this handle the situation where the separator appears in one of the key/value?

Comment (lcspinter) [reply ql/src/java/org/apache/hadoop/hive/ql/ddl/table/AbstractAlterTableOperation.java:150]: I've added a separate check to look for the separator among the property keys and throw an exception if an invalid character is found.

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java [hunk 2]
```diff
@@ -48,6 +48,9 @@
   public List<String> allowedAlterTypes = ImmutableList.of("ADDPROPS", "DROPPROPS");
   String ALTERLOCATION = "ALTERLOCATION";
   String ALLOW_PARTITION_KEY_CHANGE = "allow_partition_key_change";
+  String SET_PROPERTIES = "set_properties";
+  String UNSET_PROPERTIES = "unset_properties";
+  String PROPERTIES_SEPARATOR = "\n";
```
Comment (marton-bod) [inline standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java:53]: Not sure which characters are unsupported in a key, but I think ' might work better as a separator due to the rules of the query language (since ' is used to mark the end of the table props in the SQL, e.g. 'key'='value')

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java [hunk 3]
```diff
@@ -535,6 +538,38 @@ public void testCreateTableWithoutColumnComments() {
...
 
+  @Test
+  public void testAlterTableProperties() {
+    TableIdentifier identifier = TableIdentifier.of("default", "customers");
+    shell.executeStatement("CREATE EXTERNAL TABLE customers (" +
+        "t_int INT,  " +
+        "t_string STRING) " +
+        "STORED BY 'org.apache.iceberg.mr.hive.HiveIcebergStorageHandler' " +
+        testTables.locationForCreateTableSQL(identifier) +
+        testTables.propertiesForCreateTableSQL(ImmutableMap.of()));
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:549]: why is this part needed here?

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java [hunk 4]
```diff
@@ -535,6 +538,38 @@ public void testCreateTableWithoutColumnComments() {
...
+    String propKey = "dummy";
+    String propValue = "dummy_val";
+    shell.executeStatement(String.format("ALTER TABLE customers SET TBLPROPERTIES('%s'='%s')", propKey, propValue));
+    // Check the Iceberg table parameters
+    Table icebergTable = testTables.loadTable(identifier);
+    Assert.assertTrue(icebergTable.properties().containsKey(propKey));
+    Assert.assertEquals(icebergTable.properties().get(propKey), propValue);
+    propValue = "new_dummy_val";
+    shell.executeStatement(String.format("ALTER TABLE customers SET TBLPROPERTIES('%s'='%s')", propKey, propValue));
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:558]: nit: for readability, would it make sense to separate the 3 sections with a comment, e.g.
`// adding a new property`, 
`// changing the value of an existing property`, 
`// removing/unsetting an existing prop` ..., etc.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java [hunk 5]
```diff
@@ -29,10 +29,13 @@
 import java.util.Properties;
 import java.util.Set;
 import java.util.stream.Collectors;
+
+import org.apache.commons.lang3.exception.ExceptionUtils;
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:33]: is this a leftover import?

File: standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java (hunk 2 above)
Comment (lcspinter) [reply standalone-metastore/metastore-common/src/main/java/org/apache/hadoop/hive/metastore/HiveMetaHook.java:53]: I like that. Adding `'` to the property key, even if it's escaped, will result in an exception during query parsing.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 3 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:549]: The location is needed for HadoopTable tests. The properties section is needed to set the catalog type correctly.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 4 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:558]: Fixed.

File: iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java (hunk 5 above)
Comment (lcspinter) [reply iceberg/iceberg-handler/src/test/java/org/apache/iceberg/mr/hive/TestHiveIcebergStorageHandlerNoScan.java:33]: yes.

File: iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java [hunk 6]
```diff
@@ -251,6 +254,21 @@ public void commitAlterTable(org.apache.hadoop.hive.metastore.api.Table hmsTable
...
+      Map<String, String> contextProperties = context.getProperties();
+      if (contextProperties.containsKey(ALTER_TABLE_OPERATION_TYPE) &&
+          allowedAlterTypes.contains(contextProperties.get(ALTER_TABLE_OPERATION_TYPE))) {
+        Map<String, String> hmsTableParameters = hmsTable.getParameters();
+        Splitter splitter = Splitter.on(PROPERTIES_SEPARATOR);
+        UpdateProperties icebergUpdateProperties = icebergTable.updateProperties();
+        if (contextProperties.containsKey(SET_PROPERTIES)) {
+          splitter.splitToList(contextProperties.get(SET_PROPERTIES))
+              .forEach(k -> icebergUpdateProperties.set(k, hmsTableParameters.get(k)));
+        } else {
```
Comment (marton-bod) [inline iceberg/iceberg-handler/src/main/java/org/apache/iceberg/mr/hive/HiveIcebergMetaHook.java:267]: maybe make this `else if (contextProperties.containsKey(UNSET_PROPERTIES))`? We might later call this method from other alter table ops (e.g. schema change) and we wouldn't want to enter this block then

**Review by marton-bod — APPROVED** [review]

**Review by pvary — APPROVED** [review]
